File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/PromotionRun.java
package net.nemerosa.ontrack.model.structure;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.fasterxml.jackson.annotation.JsonView;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;
import net.nemerosa.ontrack.model.buildfilter.BuildDiff;

@Data
@AllArgsConstructor(access = AccessLevel.PROTECTED)
@JsonPropertyOrder(alphabetic = true)
public class PromotionRun implements ProjectEntity {

    private final ID id;
    @JsonView({ProjectStatusView.class, BranchStatusView.class, PromotionView.class, PromotionRun.class, Build.class, PromotionRunView.class})
    private final Build build;
    @JsonView({Build.class, PromotionRun.class, BranchBuildView.class, BuildDiff.class, BuildView.class, Decoration.class})
    private final PromotionLevel promotionLevel;
    private final Signature signature;
    private final String description;

    public static PromotionRun of(Build build, PromotionLevel promotionLevel, Signature signature, String description) {
        return new PromotionRun(
                ID.NONE,
                build,
                promotionLevel,
                signature,
                description
        );
    }

    @Override
    public Project getProject() {
        return getBuild().getProject();
    }

    @Override
    public ProjectEntityType getProjectEntityType() {
        return ProjectEntityType.PROMOTION_RUN;
    }

    @Override
    public String getEntityDisplayName() {
        return String.format("Promotion run %s/%s/%s/%s",
                build.getBranch().getProject().getName(),
                build.getBranch().getName(),
                build.getName(),
                promotionLevel.getName()
        );
    }

    public PromotionRun withId(ID id) {
        return new PromotionRun(
                id,
                build,
                promotionLevel,
                signature,
                description
        );
    }

}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/PromotionRun.kt
package net.nemerosa.ontrack.model.structure

import com.fasterxml.jackson.annotation.JsonPropertyOrder
import com.fasterxml.jackson.annotation.JsonView

@JsonPropertyOrder(alphabetic = true)
data class PromotionRun(
    override val id: ID,
    @JsonView(
        Project::class,
        PromotionLevel::class,
        PromotionRun::class,
        Build::class,
    )
    val build: Build,
    @JsonView(
        Build::class,
        PromotionRun::class,
    )
    val promotionLevel: PromotionLevel,
    override val signature: Signature,
    override val description: String?
) : ProjectEntity {

    override val project: Project
        get() = build.project

    override val parent: ProjectEntity? get() = promotionLevel

    override val projectEntityType: ProjectEntityType = ProjectEntityType.PROMOTION_RUN

    override val entityDisplayName: String
        get() = "Promotion run ${build.branch.project.name}/${build.branch.name}/${build.name}/${promotionLevel.name}"

    fun withId(id: ID) = PromotionRun(id, build, promotionLevel, signature, description)

    companion object {
        @JvmStatic
        fun of(build: Build, promotionLevel: PromotionLevel, signature: Signature, description: String?): PromotionRun =
            PromotionRun(
                ID.NONE,
                build,
                promotionLevel,
                signature,
                description
            )
    }


}
